### PR TITLE
Announce deleted devices explicitly over federation.

### PR DESCRIPTION
**Pull request by ara4n**
Adds a new `deleted` flag on the `m.device_list_update` EDU contents to notify when a device has been deleted.  This should be nicely backwards compatible with existing servers.

Specced at the bottom of https://docs.google.com/document/d/1fNBZUeMlp0fn0en5bCji5fn6mSvj48UylWfGKrk8ZIw/edit#

Fixes https://github.com/vector-im/riot-web/issues/4527
Sytest: https://github.com/matrix-org/sytest/pull/463

### CODE REVIEW

**Comment by ara4n** [comment]
@matrixbot retest this please

**Comment by richvdh** [comment]
(this is also blocked on getting some agreement on exactly what the format of the EDUs on the wire is)

File: synapse/storage/end_to_end_keys.py [hunk 1]
```diff
@@ -64,12 +64,17 @@ def _set_e2e_device_keys_txn(txn):
...
+        self, query_list, include_all_devices=False,
+        include_deleted_devices=False
+    ):
         """Fetch a list of device keys.
         Args:
             query_list(list): List of pairs of user_ids and device_ids.
             include_all_devices (bool): whether to include entries for devices
                 that don't have device keys
+            include_deleted_devices (bool): whether to include null entries for
+                devices which no longer exist (but were in the query_list)
```
[the remark this answers is not in the document]
Comment (ara4n) [reply synapse/storage/end_to_end_keys.py]: done.

File: synapse/storage/devices.py [hunk 2]
```diff
@@ -248,17 +248,32 @@ def update_remote_device_list_cache_entry(self, user_id, device_id, content,
...
+            self._simple_delete_txn(
+                txn,
+                table="device_lists_remote_cache",
+                keyvalues={
+                    "user_id": user_id,
+                    "device_id": device_id,
+                },
+            )
+
+            # Do we need this?
```
[the remark this answers is not in the document]
Comment (ara4n) [reply synapse/storage/devices.py]: i believe it is needed, as that cache is used to track device existence and we're deleting a device here. whereas in the upsert case we're not deleting a device.

**Comment by ara4n** [comment]
looks like i completely misread the code of how the EDU payloads were structured; my bad. Thanks for digging in and fixing it.

ptal?

File: synapse/storage/devices.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/storage/devices.py]: ok, the critical piece of information here was that `device_id_exists_cache` only caches the presence of devices, not the absence of them, which is why we don't need to update it when we insert a new device.

File: synapse/storage/end_to_end_keys.py [hunk 3]
```diff
@@ -64,12 +64,18 @@ def _set_e2e_device_keys_txn(txn):
         )
 
     @defer.inlineCallbacks
-    def get_e2e_device_keys(self, query_list, include_all_devices=False):
+    def get_e2e_device_keys(
+        self, query_list, include_all_devices=False,
+        include_deleted_devices=False
```
Comment (richvdh) [inline synapse/storage/end_to_end_keys.py]: trailing comma would be nice here

**Pull request merged**